File: src-tauri/src/memory/parser.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "camelCase")]
pub enum MemoryTopic {
    Profile,
    Projects,
    Rules,
    Tools,
    Writing,
    ActivityLog,
    Overrides,
    Sources,
    StaleRisks,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryEntry {
    pub id: String,
    pub topic: MemoryTopic,
    pub related_topics: Vec<MemoryTopic>,
    pub title: String,
    pub summary: String,
    pub search_text: String,
    pub source_path: String,
    pub start_line: usize,
    pub end_line: usize,
}

pub fn parse_entries(relative_path: &str, text: &str) -> Vec<MemoryEntry> {
    let mut entries = Vec::new();
    let mut current_title = String::from("Document");
    let mut current_start = 1usize;
    let mut current_lines = Vec::new();

    for (idx, line) in text.lines().enumerate() {
        let line_no = idx + 1;
        if should_split_heading(relative_path, line) {
            flush_entry(
                relative_path,
                &current_title,
                current_start,
                line_no.saturating_sub(1),
                &current_lines,
                &mut entries,
            );
            current_title = line.trim_start_matches('#').trim().to_string();
            current_start = line_no;
            current_lines.clear();
        }
        current_lines.push(line.to_string());
    }

    flush_entry(
        relative_path,
        &current_title,
        current_start,
        text.lines().count().max(current_start),
        &current_lines,
        &mut entries,
    );
    entries
}

fn should_split_heading(relative_path: &str, line: &str) -> bool {
    line.starts_with("# ")
        || line.starts_with("## ")
        || (relative_path == "memory_summary.md"
            && (line.starts_with("### ") || line.starts_with("#### ")))
}

fn flush_entry(
    relative_path: &str,
    title: &str,
    start_line: usize,
    end_line: usize,
    lines: &[String],
    out: &mut Vec<MemoryEntry>,
) {
    let body = lines.join("\n");
    let has_content = body
        .lines()
        .map(str::trim)
        .any(|line| !line.is_empty() && !line.starts_with('#'));
    if !has_content {
        return;
    }
    let summary = body
        .lines()
        .map(str::trim)
        .find(|line| {
            !line.is_empty()
                && !line.starts_with('#')
                && !line.eq_ignore_ascii_case("memory update request:")
        })
        .unwrap_or(title)
        .trim()
        .trim_start_matches("- ")
        .chars()
        .take(220)
        .collect::<String>();
    if summary.trim().is_empty() {
        return;
    }
    if is_version_preamble(title, &summary) {
        return;
    }

    let topic = infer_topic(relative_path, title, &body);
    let related_topics = infer_related_topics(&topic, title, &body);

    out.push(MemoryEntry {
        id: format!("{}:{}-{}", relative_path, start_line, end_line),
        topic,
        related_topics,
        title: title.to_string(),
        summary,
        search_text: body,
        source_path: relative_path.to_string(),
        start_line,
        end_line,
    });
}

fn is_version_preamble(title: &str, summary: &str) -> bool {
    title == "Document"
        && summary
            .strip_prefix('v')
            .is_some_and(|version| version.chars().all(|ch| ch.is_ascii_digit()))
}

fn infer_related_topics(primary_topic: &MemoryTopic, title: &str, body: &str) -> Vec<MemoryTopic> {
    if primary_topic != &MemoryTopic::Overrides {
        return Vec::new();
    }

    infer_content_topics(title, body)
}

fn infer_topic(path: &str, title: &str, body: &str) -> MemoryTopic {
    let text = format!("{} {} {}", path, title, body).to_lowercase();
    if path == "raw_memories.md"
        || path.contains("extensions/chronicle/resources")
        || path.contains("rollout_summaries/")
    {
        MemoryTopic::ActivityLog
    } else if path.contains("ad_hoc/notes") || text.contains("memory update request") {
        MemoryTopic::Overrides
    } else {
        infer_content_topics(title, body)
            .into_iter()
            .next()
            .unwrap_or(MemoryTopic::Sources)
    }
}

fn infer_content_topics(title: &str, body: &str) -> Vec<MemoryTopic> {
    let text = format!("{} {}", title, body).to_lowercase();
    let mut topics = Vec::new();

    if text.contains("user profile")
        || text.contains("技术栈")
        || text.contains("primary technical stack")
    {
        topics.push(MemoryTopic::Profile);
    }
    if text.contains("project")
        || text.contains("agent-memory-manager")
        || text.contains("beebotos")
        || text.contains("sub2api")
        || text.contains("dilidili")
    {
        topics.push(MemoryTopic::Projects);
    }
    if text.contains("preference")
        || text.contains("规则")
        || text.contains("中文输出")
        || text.contains("tool")
    {
        topics.push(MemoryTopic::Rules);
    }
    if text.contains("codex")
        || text.contains("mcp")
        || text.contains("skills")
        || text.contains("openai")
    {
        topics.push(MemoryTopic::Tools);
    }
    if text.contains("writing") || text.contains("公众号") || text.contains("写作") {
        topics.push(MemoryTopic::Writing);
    }

    topics
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classifies_profile_stack_entries() {
        let entries = parse_entries(
            "extensions/ad_hoc/notes/profile.md",
            "Memory update request:\n\n- The user's primary technical stack has shifted to Python/Rust.\n",
        );

        assert!(entries
            .iter()
            .any(|entry| entry.topic == MemoryTopic::Overrides));
        assert!(entries
            .iter()
            .any(|entry| entry.summary.contains("Python/Rust")));
        assert!(entries
            .iter()
            .any(|entry| entry.related_topics.contains(&MemoryTopic::Profile)));
    }

    #[test]
    fn keeps_full_search_text_for_multi_bullet_notes() {
        let entries = parse_entries(
            "extensions/ad_hoc/notes/profile.md",
            "Memory update request:\n\n- `dilidili` is no longer active.\n- The user's primary technical stack has shifted to Python/Rust.\n",
        );

        let entry = entries.first().expect("expected one note entry");

        assert!(entry.summary.contains("dilidili"));
        assert!(entry.search_text.contains("Python/Rust"));
        assert!(entry.related_topics.contains(&MemoryTopic::Projects));
        assert!(entry.related_topics.contains(&MemoryTopic::Profile));
    }

    #[test]
    fn keeps_chronicle_activity_out_of_profile() {
        let entries = parse_entries(
            "extensions/chronicle/resources/2026-06-02T09-09-00-10min-memory-summary.md",
            "## Memory summary\n\nThe user is reviewing BeeBotOS and current 技术栈 context in a 10 minute recording.\n",
        );

        assert!(entries
            .iter()
            .all(|entry| entry.topic == MemoryTopic::ActivityLog));
    }

    #[test]
    fn keeps_rollout_history_out_of_projects() {
        let entries = parse_entries(
            "rollout_summaries/2026-06-02T09-09-00-example.md",
            "## Task Group\n\nThe user reviewed the BeeBotOS project during this past session.\n",
        );

        assert!(entries
            .iter()
            .all(|entry| entry.topic == MemoryTopic::ActivityLog));
    }

    #[test]
    fn keeps_raw_memories_out_of_current_topics() {
        let entries = parse_entries(
            "raw_memories.md",
            "# Raw Memories\n\n## Thread `example`\n\n### Task 1: Review BeeBotOS and Codex memory\n\nThe user discussed project preferences and 技术栈 history.\n",
        );

        assert!(entries
            .iter()
            .all(|entry| entry.topic == MemoryTopic::ActivityLog));
    }

    #[test]
    fn keeps_historical_memory_update_text_out_of_corrections() {
        let entries = parse_entries(
            "raw_memories.md",
            "# Raw Memories\n\n## Thread `example`\n\nMemory update request:\n\n- The user discussed a past correction workflow.\n",
        );

        assert!(entries
            .iter()
            .all(|entry| entry.topic == MemoryTopic::ActivityLog));
    }

    #[test]
    fn skips_memory_summary_version_preamble() {
        let entries = parse_entries(
            "memory_summary.md",
            "v1\n\n## User Profile\n\nThe user's current technical stack is Python/Rust.\n",
        );

        assert!(entries.iter().all(|entry| entry.summary != "v1"));
        assert!(entries.iter().any(|entry| entry.title == "User Profile"));
    }

    #[test]
    fn splits_memory_summary_project_sections() {
        let entries = parse_entries(
            "memory_summary.md",
            "## What's in Memory\n\n### /Users/qsh/Documents/work/agent-memory-manager\n\n#### 2026-06-08\n\n- Codex memory manager MVP: agent-memory-manager, Knowledge Board, safe write\n  - desc: Iterate the memory manager UI.\n",
        );

        assert!(!entries
            .iter()
            .any(|entry| entry.title == "What's in Memory"));
        assert!(!entries
            .iter()
            .any(|entry| entry.title == "/Users/qsh/Documents/work/agent-memory-manager"));
        assert!(entries.iter().any(|entry| {
            entry.title == "2026-06-08"
                && entry.summary.contains("Codex memory manager MVP")
                && entry.topic == MemoryTopic::Projects
        }));
    }
}

File: electron/main/services/memory/parser.ts
import type { MemoryEntry, MemoryTopic } from "../../../../src/lib/types";
import { sha256, textLines } from "../shared";

export function parseEntries(relativePath: string, text: string) {
  const entries: MemoryEntry[] = [];
  const lines = textLines(text);
  let currentTitle = "Document";
  let currentStart = 1;
  let currentLines: string[] = [];

  for (const [index, line] of lines.entries()) {
    const lineNumber = index + 1;
    if (shouldSplitHeading(relativePath, line)) {
      flushEntry(relativePath, currentTitle, currentStart, Math.max(0, lineNumber - 1), currentLines, entries);
      currentTitle = line.replace(/^#+/, "").trim();
      currentStart = lineNumber;
      currentLines = [];
    }
    currentLines.push(line);
  }
  flushEntry(
    relativePath,
    currentTitle,
    currentStart,
    Math.max(lines.length, currentStart),
    currentLines,
    entries,
  );
  return entries;
}

function shouldSplitHeading(relativePath: string, line: string) {
  return (
    line.startsWith("# ") ||
    line.startsWith("## ") ||
    (relativePath === "memory_summary.md" && (line.startsWith("### ") || line.startsWith("#### ")))
  );
}

function flushEntry(
  relativePath: string,
  title: string,
  startLine: number,
  endLine: number,
  lines: string[],
  output: MemoryEntry[],
) {
  const body = lines.join("\n");
  if (!lines.some((line) => line.trim() && !line.trim().startsWith("#"))) {
    return;
  }
  const summarySource = selectSummaryLine(body) ?? (isMetadataOnlyEntry(body) ? null : title);
  if (!summarySource) {
    return;
  }
  const summary = [...summarySource].slice(0, 220).join("");
  if (!summary.trim() || (title === "Document" && /^v\d+$/.test(summary))) {
    return;
  }
  const topic = inferTopic(relativePath, title, body);
  const change = parseChangeMetadata(body);
  output.push({
    id: `${relativePath}:${sha256(`${title}\n${body}`).slice(0, 16)}`,
    topic,
    relatedTopics: topic === "overrides" ? inferContentTopics(title, body) : [],
    title,
    summary,
    searchText: body,
    sourcePath: relativePath,
    startLine,
    endLine,
    ...(change ? { change } : {}),
  });
}

function parseChangeMetadata(body: string) {
  const match = body.match(/<!--\s*agent-backplane-change\s+({[^\n]*})\s*-->/);
  if (!match) return undefined;
  try {
    const value = JSON.parse(match[1]) as Record<string, unknown>;
    if (
      typeof value.id !== "string" ||
      !["replace", "append", "revert"].includes(String(value.operation)) ||
      !Array.isArray(value.targetEntryIds) ||
      !value.targetEntryIds.every((item) => typeof item === "string") ||
      !(value.revertsChangeId === null || typeof value.revertsChangeId === "string") ||
      typeof value.createdAt !== "string"
    ) {
      return undefined;
    }
    return {
      id: value.id,
      operation: value.operation as "replace" | "append" | "revert",
      targetEntryIds: value.targetEntryIds as string[],
      revertsChangeId: value.revertsChangeId as string | null,
      createdAt: value.createdAt,
    };
  } catch {
    return undefined;
  }
}

function selectSummaryLine(body: string) {
  let inMetadataBlock = false;
  for (const line of textLines(body)) {
    const trimmed = line.trim();
    if (trimmed.startsWith("### ")) {
      inMetadataBlock = isMetadataHeading(trimmed);
      continue;
    }
    if (inMetadataBlock) {
      continue;
    }
    const normalized = normalizeSummaryLine(trimmed);
    if (isUsableSummaryLine(normalized)) {
      return normalized;
    }
  }
  return null;
}

function isMetadataOnlyEntry(body: string) {
  let inMetadataBlock = false;
  let sawNonHeading = false;
  for (const line of textLines(body)) {
    const trimmed = line.trim();
    if (trimmed.startsWith("### ")) {
      inMetadataBlock = isMetadataHeading(trimmed);
      continue;
    }
    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }
    sawNonHeading = true;
    if (!inMetadataBlock && !isRegistryMetadataLine(normalizeSummaryLine(trimmed))) {
      return false;
    }
  }
  return sawNonHeading;
}

function normalizeSummaryLine(line: string) {
  let value = line.trim();
  for (const prefix of ["- ", "desc:", "learnings:"]) {
    if (value.startsWith(prefix)) {
      value = value.slice(prefix.length).trim();
    }
  }
  return value;
}

function isUsableSummaryLine(line: string) {
  return Boolean(
    line &&
      !line.startsWith("#") &&
      line.toLowerCase() !== "memory update request:" &&
      line !== "§" &&
      !line.startsWith("<!-- agent-backplane-change") &&
      !isRegistryMetadataLine(line),
  );
}

function isMetadataHeading(line: string) {
  const heading = line.replace(/^#+/, "").trim().toLowerCase();
  return heading === "rollout_summary_files" || heading === "keywords";
}

function isRegistryMetadataLine(line: string) {
  const lower = line.toLowerCase();
  return (
    lower.startsWith("scope:") ||
    lower.startsWith("applies_to:") ||
    lower.startsWith("rollout_summaries/") ||
    lower.includes("rollout_path=") ||
    lower.includes("thread_id=")
  );
}

function inferTopic(path: string, title: string, body: string): MemoryTopic {
  const text = `${path} ${title} ${body}`.toLowerCase();
  if (
    path === "raw_memories.md" ||
    path.includes("extensions/chronicle/resources") ||
    path.includes("rollout_summaries/")
  ) {
    return "activityLog";
  }
  if (path.includes("ad_hoc/notes") || text.includes("memory update request")) {
    return "overrides";
  }
  return inferContentTopics(title, body)[0] ?? "sources";
}

function inferContentTopics(title: string, body: string) {
  const text = `${title} ${body}`.toLowerCase();
  const topics: MemoryTopic[] = [];
  if (text.includes("user profile") || text.includes("技术栈") || text.includes("primary technical stack")) {
    topics.push("profile");
  }
  if (
    text.includes("project") ||
    text.includes("agent-backplane") ||
    text.includes("beebotos") ||
    text.includes("sub2api") ||
    text.includes("dilidili")
  ) {
    topics.push("projects");
  }
  if (text.includes("preference") || text.includes("规则") || text.includes("中文输出") || text.includes("tool")) {
    topics.push("rules");
  }
  if (text.includes("codex") || text.includes("mcp") || text.includes("skills") || text.includes("openai")) {
    topics.push("tools");
  }
  if (text.includes("writing") || text.includes("公众号") || text.includes("写作")) {
    topics.push("writing");
  }
  return topics;
}
